aggregator/src/aggregator/key_rotator.rs

```diff
@@ -0,0 +1,911 @@
+///
+/// # Global HPKE Keys
+///
+/// The key rotator can handle key rotation for global HPKE keys. The key rotator is tolerant of
+/// some manual operator changes to the `global_hpke_keys` table.
+///
+/// Operators _must not_ manually change the `created_at` or `updated_at` columns, as these are
+/// used for rotation decision-making.
+///
+/// It is strongly discouraged to:
+///   - Delete active or expired keypairs too early, as that would leave Janus unable to decrypt
+///     report shares using the keypair.
+///   - Promote pending keypairs too early, unless all Janus replicas have been rebooted since
+///     their insertion.
+///   - Directly insert active keypairs.
+///
+/// Keypairs that are manually inserted are adopted by the key rotator and will have their lifecycle
+/// managed. The number of keypairs per ciphersuite will trend towards 1, i.e. 2 keys with the same
+/// ciphersuite will eventually be replaced with 1 key.
+#[derive(Debug)]
+pub struct KeyRotator<C: Clock> {
+    datastore: Arc<Datastore<C>>,
+    hpke: Vec<HpkeKeyRotatorConfig>,
+}
+
+/// Defines the ciphersuite and rotation policy of a global HPKE key.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct HpkeKeyRotatorConfig {
+    /// How long key remains in [`HpkeKeyState::Pending`] before being moved to
+    /// [`HpkeKeyState::Active`]. This should be greater than
+    /// [`AggregatorConfig::global_hpke_configs_refresh_interval`].
+    #[serde(rename = "pending_duration_secs", default = "default_pending_duration")]
+    pub pending_duration: Duration,
+
+    /// The time-to-live of the key. Once this is exceeded, the key is moved to
+    /// [`HpkeKeyState::Expired`]. It is at operator discretion as to how long this should be.
+    #[serde(rename = "active_duration_secs", default = "default_active_duration")]
+    pub active_duration: Duration,
+
+    /// How long the key remains in [`HpkeKeyState::Expired`] before being deleted. This should
+    /// be greater than the clients' HPKE key cache maximum age.
+    #[serde(rename = "expired_duration_secs", default = "default_expired_duration")]
+    pub expired_duration: Duration,
+
+    /// The ciphersuite of the key.
+    #[serde(default = "default_hpke_ciphersuite")]
+    pub ciphersuite: HpkeCiphersuite,
+
+    /// Safely phase out this ciphersuite. This will immediately delete any pending keys and
+    /// expire any active keys. Expired keys are retained for at least [`Self::expired_duration`]
+    /// before being deleted.
+    ///
+    /// After [`Self::expired_duration`], this config can be removed entirely. It is not recommended
+    /// to remove key configurations without running the key rotator with `retire` first.
+    ///
+    /// There must be at least one non-retired config present with its key in the active state. If
+    /// this is the only config and it is marked retired, the key rotator will refuse to run.
+    #[serde(default)]
+    pub retire: bool,
```

Comment:
That SGTM.

I had introduced `retire` to work around the fact that we would need some "global" set of durations to handle unaccounted for keys, but yeah it's just as simple to have global durations for every key.

Reply:
Good suggestion, this should be operationally simpler.

I'm treating "unknown" keys, i.e. ones that we don't have a ciphersuite in the config for, as ones that should be imminently retired. I figure this case is most often the operator removing a ciphersuite entry from the config.

It does mean that any keys that are manually inserted by the operator and aren't in the list of known ciphersuites are unceremoniously expired/deleted though. I'd prefer this so that we retain the predictable behavior of config removal resulting in key retirement.